decisions/0014-context-middleware.md

```diff
@@ -0,0 +1,254 @@
+# Middleware + Context
+
+Date: 2025-01-22
+
+Status: proposed
+
+## Context
+
+_Lol "context", get it 😉_
+
+The [Middleware RFC][rfc] is the _most-upvoted_ RFC/Proposal in the React Router repo. We actually tried to build and ship it quite some time ago but realized that without single fetch it didn't make much sense in an SSR world for 2 reasons:
+
+- With the individual HTTP requests per loader, middleware wouldn't actually reduce the # of queries to your DB/API's - it would just be a code convenience with no functional impact
+- Individual HTTP requests meant a lack of a shared request scope across routes
+
+We've done a lot of work since then to get us to a place where we could ship a middleware API we were happy with:
+
+- Shipped [Single Fetch][single-fetch]
+- Shipped [`dataStrategy`][data-strategy] for DIY middleware in React Router SPAs
+- Iterated on middleware/context APIs in the [Remix the Web][remix-the-web] project
+- Developed a non-invasive type-safe + composable [context][async-provider] API
+
+## Decision
+
+### Leverage a new type-safe `context` API
+
+We originally considered leaning on our existing `context` value we pass to server-side `loader` and `action` functions, and implementing a similar client-side equivalent for parity. However, the type story around `AppLoadContext` isn't great, so that would mean implementing a new API client side that we knew we weren't happy with from day one. And then likely replacing it with a better version fairly soon after.
+
+Instead, when the flag is enabled, we'll be removing `AppLoadContext` in favor of a type-safe `context` API that is similar in usage to the `React.createContext` API:
```

Comment:
This will be a breaking change for many apps, is there no way to keep the AppLoadContext? I know many use cases can be solved by middlewares but being able to inject from the HTTP server into the RR app is useful

Reply:
From my initial testing it seems like it'll still be possible - you'll just need to pass it a bit differently 😃

Reply:
Correct - you can use `getLoadContext` to pass values through from the adapter - it just needs to conform to the new `ContextProvider` API and return a `Map<RouterContext, unknown>`: https://github.com/remix-run/react-router/blob/brophdawg11/feat-middleware/playground/middleware/server.ts#L26

Reply:
ohh in that case there's no issue, I would worry that this is a breaking change tho so should this be marked as v8_middleware once stable?

Reply:
Yeah - this is a known breaking change and the flag will stabilize as `future.v8_middleware` once it's ready